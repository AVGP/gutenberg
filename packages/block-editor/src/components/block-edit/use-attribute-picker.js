/**
 * External dependencies
 */
import { upperFirst } from 'lodash';

/**
 * WordPress dependencies
 */
import { useMemo } from '@wordpress/element';
import { useShallowCompareDep, useDeepCompareDep } from '@wordpress/compose';

/**
 * Internal dependencies
 */
import { useBlockEditContext } from './context';

/**
 * Low level hook that takes an options object that has a list of attributes to manage, values that they can be set to,
 * and functions for customizing how values are compared, saved, and returned. It also handles setting attributes to
 * custom values and provides an easy way to bind utilities.
 * It returns an object with mapped attributes, attribute setters, and bound utilities.
 * It can be used to write hooks like `useColors`.
 *
 * @see [useColors](/packages/block-editor/README.md#useColors)
 *
 * @param {Object}                 attributePickerOptions                           The options object.
 *
 * @param {string[]}               attributePickerOptions.names                     The list of attributes to manage.
 *
 * @param {*[]}                    attributePickerOptions.valuesEnum                The list of values attributes can be set to.
 *                                                                                   Setting an attribute to a custom value not on
 *                                                                                   this list will set the attribute's value to
 *                                                                                   `undefined` and the corresponding `custom${ upperFirst( name ) }`
 *                                                                                   attribute to the custom value.
 *
 * @param {Function}               [attributePickerOptions.findEnumValue]           Function for comparing an enum value to the
 *                                                                                   value passed to the setter, when looking for
 *                                                                                   the value in the enum. Defaults to shallow comparison.
 *
 * @param {Function}               [attributePickerOptions.mapEnumValue]            Function for mapping the enum value found with
 *                                                                                   `findEnumValue` before saving it, if found.
 *                                                                                   Otherwise, the value is assumed to be custom and saved as is.
 *
 * @param {*[]}                    [attributePickerOptions.findAndMapEnumValueDeps] List of items that, if they fail a shallow equality test, will
 *                                                                                   cause new setters to be returned. Useful if your `findEnumValue` or `mapEnumValue` have closures.
 *
 * @param {Function}               [attributePickerOptions.mapAttribute]            Function for mapping attributes before returning them, called
 *                                                                                   with the attribute's or custom attribute's value and the name.
 *
 * @param {*[]}                    [attributePickerOptions.mapAttributeDeps]        List of items that, if they fail a shallow equality test, will
 *                                                                                   cause attributes to be mapped again. Useful if your `mapAttribute` has closures.
 *
 * @param {{ string: Function }}   [attributePickerOptions.utilsToBind]             Object with functions to bind to `valuesEnum` and merge into the
 *                                                                                   returned object. Useful for utilities that need to be bound.
 *
 * @param {*[]}                    [attributePickerOptions.utilsToBindDeps]         List of items that, if they fail a shallow equality test, will
 *                                                                                   cause `utilsToBind` to be bound again. Useful if any of your utils has closures.
 *
 * @return {Object} The object with mapped attributes and attribute setters.
 */
export default function useAttributePicker( {
	names,
	valuesEnum,

	findEnumValue = ( enumValue, newValue ) => enumValue === newValue,
	mapEnumValue = ( enumValue ) => enumValue,
	findAndMapEnumValueDeps = [],

	mapAttribute = ( mappedEnumValueOrCustomValue ) => mappedEnumValueOrCustomValue,
	mapAttributeDeps = [],

	utilsToBind = {},
	utilsToBindDeps = [],
} ) {
	const { setAttributes, attributes } = useBlockEditContext();
	const namesDep = useShallowCompareDep( names );
	const valuesEnumDep = useDeepCompareDep( valuesEnum );

	const setters = useMemo(
		() =>
			names.map( ( name ) => ( newValue ) => {
				const foundEnumValue = valuesEnum.find( ( enumValue ) =>
					findEnumValue( enumValue, newValue )
				);
				const mappedValue = foundEnumValue && mapEnumValue( foundEnumValue );

				setAttributes( {
					[ name ]: mappedValue ? mappedValue : undefined,
					[ `custom${ upperFirst( name ) }` ]: mappedValue ? undefined : newValue,
				} );
			} ),
		[ namesDep, valuesEnumDep, ...findAndMapEnumValueDeps ]
	);

	const mappedAttributes = useMemo(
		() =>
			names.map( ( name ) =>
				mapAttribute(
					attributes[ name ] || attributes[ `custom${ upperFirst( name ) }` ],
					name
				)
			),
		[
			namesDep,
			valuesEnumDep,
			useDeepCompareDep(
				names.map(
					( name ) => attributes[ name ] || attributes[ `custom${ upperFirst( name ) }` ]
				)
			),
			...mapAttributeDeps,
		]
	);

	const utils = useMemo(
		() =>
			Object.keys( utilsToBind ).reduce( ( utilsAccumulator, utilName ) => {
				utilsAccumulator[ utilName ] = utilsToBind[ utilName ].bind(
					null,
					valuesEnum
				);
				return utilsAccumulator;
			}, {} ),
		[ valuesEnumDep, useShallowCompareDep( utilsToBind ), ...utilsToBindDeps ]
	);

	return useMemo(
		() => ( {
			...names.reduce( ( attributePicker, name, i ) => {
				attributePicker[ `set${ upperFirst( name ) }` ] = setters[ i ];
				attributePicker[ name ] = mappedAttributes[ i ];
				return attributePicker;
			}, {} ),
			utils,
		} ),
		[ setters, mappedAttributes, utils ]
	);
}
