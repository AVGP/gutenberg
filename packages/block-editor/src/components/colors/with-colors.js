/**
 * WordPress dependencies
 */
import { createHigherOrderComponentWithMergeProps } from '@wordpress/compose';

/**
 * Internal dependencies
 */
import useColors, { useCustomColors } from './use-colors';

/**
 * A higher-order component factory for creating a 'withCustomColors' HOC, which handles color logic
 * for class generation color value, retrieval and color attribute setting.
 *
 * Use this higher-order component to work with a custom set of colors.
 *
 * @example
 *
 * ```jsx
 * const CUSTOM_COLORS = [ { name: 'Red', slug: 'red', color: '#ff0000' }, { name: 'Blue', slug: 'blue', color: '#0000ff' } ];
 * const withCustomColors = createCustomColorsHOC( CUSTOM_COLORS );
 * // ...
 * export default compose(
 *     withCustomColors( 'backgroundColor', 'borderColor' ),
 *     MyColorfulComponent,
 * );
 * ```
 *
 * @param {Array} colorsArray The array of color objects (name, slug, color, etc... ).
 *
 * @return {Function} Higher-order component.
 */
export function createCustomColorsHOC( colorsArray ) {
	return ( ...colorTypes ) =>
		createHigherOrderComponentWithMergeProps(
			() => useCustomColors( colorTypes, colorsArray ),
			'withCustomColors'
		);
}

/**
 * A higher-order component, which handles color logic for class generation color value, retrieval and color attribute setting.
 *
 * For use with the default editor/theme color palette.
 *
 * @example
 *
 * ```jsx
 * export default compose(
 *     withColors( 'backgroundColor', { textColor: 'color' } ),
 *     MyColorfulComponent,
 * );
 * ```
 *
 * @param {...(object|string)} colorTypes The arguments can be strings or objects. If the argument is an object,
 *                                        it should contain the color attribute name as key and the color context as value.
 *                                        If the argument is a string the value should be the color attribute name,
 *                                        the color context is computed by applying a kebab case transform to the value.
 *                                        Color context represents the context/place where the color is going to be used.
 *                                        The class name of the color is generated using 'has' followed by the color name
 *                                        and ending with the color context all in kebab case e.g: has-green-background-color.
 *
 * @return {Function} Higher-order component.
 */
export default function withColors( ...colorTypes ) {
	return createHigherOrderComponentWithMergeProps(
		() => useColors( colorTypes ),
		'withColors'
	);
}
