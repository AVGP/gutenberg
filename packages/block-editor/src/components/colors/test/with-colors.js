/**
 * External dependencies
 */
import { mount } from 'enzyme';

/**
 * Internal dependencies
 */
import { BlockEditContextProvider } from '../../block-edit';
import { createCustomColorsHOC } from '../with-colors';

describe( 'createCustomColorsHOC', () => {
	it( 'provides the the wrapped component with color values and setter functions as props', () => {
		const withCustomColors = createCustomColorsHOC( [ { name: 'Red', slug: 'red', color: 'ff0000' } ] );
		const EnhancedComponent = withCustomColors( 'backgroundColor' )( () => (
			<div />
		) );

		const wrapper = mount(
			<BlockEditContextProvider
				value={ { attributes: { backgroundColor: null } } }
			>
				<EnhancedComponent />
			</BlockEditContextProvider>
		);

		expect( wrapper ).toMatchSnapshot();
	} );

	it( 'setting the color to a value in the provided custom color array updated the backgroundColor attribute', () => {
		const withCustomColors = createCustomColorsHOC( [ { name: 'Red', slug: 'red', color: 'ff0000' } ] );
		const EnhancedComponent = withCustomColors( 'backgroundColor' )( ( props ) => (
			<button onClick={ () => props.setBackgroundColor( 'ff0000' ) }>Test Me</button>
		) );

		const setAttributes = jest.fn();

		const wrapper = mount(
			<BlockEditContextProvider
				value={ { attributes: { backgroundColor: null }, setAttributes } }
			>
				<EnhancedComponent />
			</BlockEditContextProvider>
		);

		wrapper.find( 'button' ).simulate( 'click' );
		expect( setAttributes ).toHaveBeenCalledWith( { backgroundColor: 'red', customBackgroundColor: undefined } );
	} );

	it( 'setting the color to a value not in the provided custom color array updates customBackgroundColor attribute', () => {
		const withCustomColors = createCustomColorsHOC( [ { name: 'Red', slug: 'red', color: 'ff0000' } ] );
		const EnhancedComponent = withCustomColors( 'backgroundColor' )( ( props ) => (
			<button onClick={ () => props.setBackgroundColor( '000000' ) }>Test Me</button>
		) );

		const setAttributes = jest.fn();

		const wrapper = mount(
			<BlockEditContextProvider
				value={ { attributes: { backgroundColor: null }, setAttributes } }
			>
				<EnhancedComponent />
			</BlockEditContextProvider>
		);

		wrapper.find( 'button' ).simulate( 'click' );
		expect( setAttributes ).toHaveBeenCalledWith( { backgroundColor: undefined, customBackgroundColor: '000000' } );
	} );
} );
