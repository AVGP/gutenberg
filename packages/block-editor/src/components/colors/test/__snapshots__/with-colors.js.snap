// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createCustomColorsHOC provides the the wrapped component with color values and setter functions as props 1`] = `
<WithCustomColors(Component)>
  <Component>
    <Component
      backgroundColor={
        Object {
          "class": undefined,
          "color": undefined,
        }
      }
      colorUtils={
        Object {
          "getMostReadableColor": [Function],
        }
      }
      setBackgroundColor={[Function]}
    >
      <div />
    </Component>
  </Component>
</WithCustomColors(Component)>
`;
